src/app/view/properties/still-image-osdviewer/still-image-osdviewer.component.ts

```diff
@@ -34,86 +41,273 @@ declare var OpenSeadragon: any;
+    let fileValues : ReadStillImageFileValue[] = [];
+    for (let resource of this.resources) {
+      let fileValuesResource = resource.properties[AppConfig.hasStillImageFileValue];
+      Array.prototype.push.apply(fileValues, fileValuesResource);
     }
+    return fileValues.filter( (image) => { return !image.isPreview; } );
+  }
 
-    // If this.ignorePreviews == true:
-    //  returns a shallow copy of the this.fileValues array, omitting all entries with entry.isPreview == true.
-    // else:
-    //  returns the this.fileValues array itself
-    private getImagesToDisplay() : ReadStillImageFileValue[] {
-      if (this.ignorePreviews) {
-        return this.fileValues.filter( (image) => { return !image.isPreview; } );
-      } else {
-        return this.fileValues;
+  /**
+  * Returns an array containing all regions of all (non-preview) images in this.resources
+  * The returneed array has the form [ [ image0_region0, image0_region1, ... ], [ image1_region0, image1_region1 ], ... ]
+  * @returns {ReadResource[][]}
+  */
+  private getRegionsToDisplay() : ReadResource[][] {
+    let regions = [];
+    for (let resource of this.resources) {
+      let incomingRegions = resource.incoming.filter((resource) => { return resource.type == AppConfig.Region});
+      regions.push(incomingRegions);
+    }
+    return regions;
+  }
+
+  /**
+  * Adds a ROI-overlay to the viewer for every region of every (non-preview) image in this.resources
+  */
+  private renderRegions() : void {
+    this.viewer.clearOverlays();
+    let regionsByResource = this.getRegionsToDisplay();
+    for (let resourceIndex in regionsByResource) {
+      let regions = regionsByResource[resourceIndex];
+      for (let region of regions) {
+        let geometries = region.properties[AppConfig.hasGeometry] as ReadGeomValue[];
+        for (let geometryValue of geometries) {
+          let geometry = geometryValue.geometry;
+          this.createSVGOverlay(geometry, Number(resourceIndex));
+        }
       }
     }
+  }
+
+  /**
+  * Creates and adds a ROI-overlay to the viewer
+  * @param {RegionGeometry} geometry - the geometry describing the ROI
+  * @param {number} resourceIndex - the index (to this.resources) of the resource/image to which the ROI belongs.
+  */
+  private createSVGOverlay(geometry: RegionGeometry, resourceIndex : number) : void {
+    let lineColor = geometry.lineColor;
+    let lineWidth = geometry.lineWidth;
+
+    let imagesToDisplay = this.getImagesToDisplay();
+    let imgWidth = imagesToDisplay[resourceIndex].dimX;
+    let imgHeight = imagesToDisplay[resourceIndex].dimY;
+    let aspectRatio = (imgHeight/imgWidth);
+
+    let svgElement;
+    switch (geometry.type) {
+      case "rectangle":
+        svgElement = document.createElementNS('http://www.w3.org/2000/svg', 'polygon');  // yes, we render rectangles as svg polygon elements
+        this.addSVGAttributesRectangle(svgElement, geometry, aspectRatio, resourceIndex);
```

Comment:
Since commit 8c9a312 we do not rely on this index any more. 
(For the sake of clarification: the index was corresponding to the x-offset of the image in the viewer. Images are rendered in a horizontal line in the viewer and have a width of 1 in the viewport coord system of OSD, so the index to the (now obsolete) `getImagesToDisplay()` array was corresponding to the x-offset of the respective image in viewport coords.
This x-offset is added to the coordinates of the geometry coordinates to get from coords relative to a image to viewport coords.
The general approach is still the same, but we are not relying directly on a array index any more. Look for xOffset and related comments in the new code.)